packages/amazon-bedrock/src/bedrock-event-stream-response-handler.ts

```diff
@@ -0,0 +1,104 @@
+import { EmptyResponseBodyError } from '@ai-sdk/provider';
+import {
+  ParseResult,
+  safeParseJSON,
+  extractResponseHeaders,
+  ResponseHandler,
+} from '@ai-sdk/provider-utils';
+import { EventStreamCodec } from '@smithy/eventstream-codec';
+import { toUtf8, fromUtf8 } from '@smithy/util-utf8';
+import { ZodSchema } from 'zod';
+
+// https://docs.aws.amazon.com/lexv2/latest/dg/event-stream-encoding.html
+export const createBedrockEventStreamResponseHandler =
+  <T>(
+    chunkSchema: ZodSchema<T>,
+  ): ResponseHandler<ReadableStream<ParseResult<T>>> =>
+  async ({ response }: { response: Response }) => {
+    const responseHeaders = extractResponseHeaders(response);
+
+    if (response.body == null) {
+      throw new EmptyResponseBodyError({});
+    }
+
+    const codec = new EventStreamCodec(toUtf8, fromUtf8);
+    let buffer = new Uint8Array(0);
+
+    return {
+      responseHeaders,
+      value: response.body.pipeThrough(
+        new TransformStream<Uint8Array, ParseResult<T>>({
+          transform(chunk, controller) {
+            // Append new chunk to buffer.
+            const newBuffer = new Uint8Array(buffer.length + chunk.length);
+            newBuffer.set(buffer);
+            newBuffer.set(chunk, buffer.length);
+            buffer = newBuffer;
+
+            // Try to decode messages from buffer.
+            while (buffer.length >= 4) {
+              // The first 4 bytes are the total length (big-endian).
+              const totalLength = new DataView(
+                buffer.buffer,
+                buffer.byteOffset,
+                buffer.byteLength,
+              ).getUint32(0, false);
+
+              // If we don't have the full message yet, wait for more chunks.
+              if (buffer.length < totalLength) {
+                break;
+              }
+
+              try {
+                // 3) Decode exactly the sub-slice for this event.
+                const subView = buffer.subarray(0, totalLength);
+                const decoded = codec.decode(subView);
+
+                // Slice the used bytes out of the buffer, removing this message.
+                buffer = buffer.slice(totalLength);
+
+                // Process the message.
+                if (decoded.headers[':message-type']?.value === 'event') {
+                  const data = new TextDecoder().decode(decoded.body);
```

Comment:
extract text decode, this is most likely a hot loop and we could init it before the loop (i.e. outside of the transform stream)